arch/arm/src/samv7/sam_qencoder.c

```diff
@@ -270,11 +316,139 @@ static int sam_reset(struct qe_lowerhalf_s *lower)
 static int sam_ioctl(struct qe_lowerhalf_s *lower, int cmd,
                      unsigned long arg)
 {
-  /* No ioctl commands supported */
-
+#ifdef CONFIG_SAMV7_QENCODER_ENABLE_GETINDEX
+  switch (cmd)
+    {
+      case QEIOC_GETINDEX:
+        {
+          /* Call the qeindex function */
+
+          sam_qeindex(lower, (struct qe_index_s *) arg);
+          return OK;
+        }
+
+      default:
+        {
+          return -ENOTTY;
+        }
+    }
+#else
   return -ENOTTY;
+#endif
 }
 
+#ifdef CONFIG_SAMV7_QENCODER_ENABLE_GETINDEX
+/****************************************************************************
+ * Name: sam_qe_pos_16to32b
+ *
+ * Description:
+ *   An inline function performing the extension of current position.
+ *   Last reading is saved to priv->last_pos.
+ *
+ ****************************************************************************/
+
+static inline int32_t sam_qe_pos_16to32b(struct qe_lowerhalf_s *lower,
+                                         uint32_t current_pos)
+{
+  struct sam_lowerhalf_s *priv = (struct sam_lowerhalf_s *) lower;
+
+  uint32_t new_pos = *(volatile uint32_t *) &priv->last_pos;
+  new_pos += (int16_t) (current_pos - new_pos);
+  *(volatile uint32_t *) &priv->last_pos = new_pos;
+
+  return (int32_t) new_pos;
+}
+#endif
+
+#ifdef CONFIG_SAMV7_QENCODER_ENABLE_GETINDEX
+/****************************************************************************
+ * Name: sam_qe_indx_pos_16to32b
+ *
+ * Description:
+ *   An inline function performing the extension of the last index position.
+ *   Last reading is saved to priv->last_index.
+ *
+ ****************************************************************************/
+
+static inline int32_t sam_qe_indx_pos_16to32b(struct qe_lowerhalf_s *lower,
+                                              uint32_t current_indx_pos)
+{
+  struct sam_lowerhalf_s *priv = (struct sam_lowerhalf_s *) lower;
+
+  uint32_t new_index = *(volatile uint32_t *) &priv->last_pos;
```

Comment:
why need cast to `volatile uint32_t *`

Reply:
The cast can be even to uint16_t but it has to be uintX_t, because modulo arithmetic is only defined for unsigned types in C language. At the moment, where two signed types are subtracted then if there has been overflow or result does not fit into target type then the behavior is undefined. On the other hand subtraction of unsigned types has precisely defined behavior. In the fact you get internally signed type which has sign as an additional bit. When size is limited then it again works with larger unsigned type to have defined behavior when overflow at or over type module (2 power to n-bits) appears or if there is underflow under zero.

If the type is not unsigned then following operation after overflow do not have defined behavior.  So that is why the sequence is  

```
  uint32_t new_index = *(volatile uint32_t *) &priv->last_pos;
  new_index += (int16_t) (current_indx_pos - new_index);
  *(volatile uint32_t *) &priv->last_index = new_index;

  return (int32_t) new_index;
```

When we consider that even for some long distance car track or some pump control for speed etc.  even extended 32-bit value can overflow then even operations in modulo should be used for controllers in user-space. So it would be better to return both, IRC value and  index as 32-bit unsigned. But because the NuttX standard is given as signed and for some mechanic with motion  within limited range, it is more natural to use +/- range around zero, we keep signed output. That is why last_pos is kept as uint32_t even that values visible to user-space are signed. Copy with volatile then ensures operation on the copy taken by single read which ensures correct operation even when code is accessed from multiple threads even on multiple CPUs in parallel.

As for the choice between unit16_t and uint32_t for values which are limited to int16_t after subtraction, on ARM there are no native operations for 16 or 8-bit wide types, so range limitations has to be introduced by additional machine operations during each arithmetic operation. Return values and arguments to the functions are passed in full 32-bit register in the most cases. So again limitation to 16-bits results in abundant instructions. They can be eliminated when static inline is used because GCC can find that these problematic bits do not contribute to the results when intermediate value is limited to int16_t. Even if we consider x86 in 32-bit and 64-bit modes, then 8-bit and 32-bit operations are natural, 16-bit ones require additional prefix and cause problematic third (or with CC fourth) input dependency to instructions. x86 64-operations require REX prefix but there is optimization for them and 32-bit ones zeros top part so no additional dependency is introduced. So in general 16-bit integers are generally problematic for most of today processors and should be used only to store results into memory if data are large and limited range is enough. FP16 and Co. and 16-bit vector elements are different story on GPUs and multimedia extensions...

Reply:
The volatile ensures that the compiler loads previous position value exactly only once from global (shared) location and proceeds both subtraction and following addition on this single copy which results in extension to 32-bits. But if compiler decides for some reason to subtract against one value and then some another thread/process updates shared value and addition is done against reload updated value then result is incorrect. So we want to ensure that the value from the global state is fetched exactly once into local copy.

Reply:
but it's unnecessary here. If you want the synchronization between thread or irq, atomic/spinlock/mutex is the right tool to use: https://www.kernel.org/doc/html/next/process/volatile-considered-harmful.html.

Reply:
The option is to protect IOCTL operations access by critical section, but it is not necessary in this case under later above and later described conditions. Code is written such way that for example high priority PMSM control task can read position by IOCT and it position can be accessed by low priority task for monitoring purposes in parallel. No locking is necessary but what is mandatory that subtraction and following addition is done against the same value last_pos snapshot inside each invocation of sam_qe_pos_16to32b(). The last_pos field can be changed to atomic and atomic access can be used but it is not necessary and barrier instructions are overhead.

There are next assumptions which allows correct operation

1. the position is read by some of the treads more often than is the time for which position changes by +/- 32k increments for maximal considered speed.
2. no thread (even some very low priority one) stays inside sam_qe_pos_16to32b() longer than is the time equivalent for +/- 32k position change

Under these assumptions there is not necessary any additional locking. If these assumptions are not met then there is problem which can be solved only in hardware, may be there is some option to cascade counters etc. 32-bit counters found on some other MCUs are even better. But I actual solution fits our needs, has minimal performance impact and I expect that it would fit 99+% of needs of others.

Reply:
I am fine with the cast, but volatile mayn't work as you expect. For example, CPU may reorder the instruction sequence just like compiler. volatile can't , but memory barrier, atomic etc could forbid CPU do this type of optimization.

Reply:
It is not problem if CPU reorders operations, there is only single situation which needs to be prevented and it is situation when the subtract is done against one snapshot of the last_pos and following addition is done against another value updated by some other thread, IRQ or anything else. The volatile modifier at pointer prevents compiler to load the value of the last_pos the second time between subtract and following addition. It is not probable on ARM where both operands have to reside in the registers before add machine code, but for example on x86 sub and add can be done against last_pos value accessed by offset addressing against some base register, so memory is accessed twice. Volatile ensures that value is read exactly once from memory and stored in the local variable (register) and all is fine whatever value is read. There can be concern if the value is read by halves or individual bytes. But that is something which is silently considered to not happen for  *(volatile uint32_t *) & combination on 32-bit target. It is right that for 16-bit CPU this could be a problem. If that is problem for 32-bit ones then you cannot use *(volatile uint32_t *) to access registers which can change their value independently from CPU action, because you cannot get consistent 32-bit data in such case. If you drop volatile in the discussed case, then probability that value in memory is accessed twice on ARM is low, zero for GCC with some optimization set and even for unoptimized case when local variable is used. Combined with probability that some other parallel thread reads and extends value at exactly same time instant and fits with update directly between subtraction and addition is again extremely low so code will work without volatile but is theoretically susceptible to misbehave. With volatile it should be OK even if multiple cores are used and CPU does not keep value in reorder registers for time longer than 32k  increments changes. But yes, on multiple cores power PC or Apple M1 when some core loops in some close busy loop (for time around 1 millisecond and more for 32 mega IRC events/s) and does not read value from the counter again, I can imagine that barrier instruction could be necessary. But on such architectures you often need to add some barrier instructions even to basic read and writes into I/O space to prevent reorder of such accesses. So it is totally another category than Cortex-M based cores.